### PR TITLE
Generalize enrichment script

### PR DESCRIPTION
**Pull request by ivagljiva**
This PR contains a generalized version of @adw96's excellent functional enrichment script so that we can compute enrichment statistics for anything we like :) 
(I am not suggesting we merge this anytime soon, I just wanted to offer it up for suggestions and feedback)

### Enrichment script
I haven't changed the actual workings of Amy's script, I only altered which input it accepts so that we can use it in more situations. The generalized version of the script is called `anvi-script-enrichment-stats`. Also, please note that I didn't actually remove `anvi-script-run-functional-enrichment-stats` from our sandbox so that people can still use it. 

The columns that this script accepts are now:

| (any header) | accession | (any header) | associated_groups | p_XX | N_XX | .... |
|---|---|---|---|---|---|---|
| (anything, but presumably will be names) | (accessions like COG0493 or K01234) | (anything, for example list of gene cluster IDs or sample IDs) | (comma-separated list of associated groups) | (proportion) | (int) |

Here's an example of what the input looks like for KEGG modules:
KEGG_MODULE | accession | sample_ids | associated_groups | p_R | N_R | p_H | N_H | p_N | N_N
---|---|---|---|---|---|---|---|---|---|
2-Oxocarboxylic acid chain extension, 2-oxoglutarate => 2-oxoadipate => 2-oxopimelate => 2-oxosuberate | M00608 | sample_80,sample_128,... | H | 0.0625 | 16 | 0.23076923076923078 | 13 | 0.0 | 8
ADP-L-glycero-D-manno-heptose biosynthesis | M00064 | sample_101,sample_27,... | H | 0.25 | 16 | 0.6153846153846154 | 13 | 0.375 | 8
Acetogen | M00618 | sample_101,sample_13,... | R,H | 0.875 | 16 | 1.0 | 13 | 0.625 | 8

### Driver program
I also wrote a driver program called `anvi-compute-enrichment-scores` that is a general version of `anvi-get-enriched-functions-per-pan-group`.
-  It has the option to do the same thing(s) as `anvi-get-enriched-functions-per-pan-group` (in fact as of right now it still relies upon the old enrichment script that is specific for pangenomics, but I will change this)
- It also provides the option to find enriched KEGG modules based on the output of `anvi-estimate-metabolism`
  - to do this, it converts the 'modules' mode output from `anvi-estimate-metabolism` into the proper input format before calling the enrichment script
  - this part relies on a new class in `kegg.py`, `KeggModuleEnrichment`
- we can extend this driver program later to allow us to compute enrichment statistics on anything else we want (ie, non-pangenome functions). All we have to do is add new input options to the driver program and make sure that those inputs are converted to the proper format for `anvi-script-enrichment-stats`

### Usage
Using `anvi-compute-enrichment-scores` to get enriched functions in pangenome groups should be exactly the same as using `anvi-get-enriched-functions-per-pan-group`, as it takes the same parameters.

But here is an example of the new stuff (enriched modules):

**Standard modules enrichment usage**

```
anvi-compute-enrichment-scores -M modules.txt -G groups.txt -o enriched_modules.txt
```
The required inputs for module enrichment are 1) the modules mode output and 2) a 2-column tab-delimited file matching the sample names from the modules mode output to the groups they belong to.

**Example modules.txt**

unique_id | db_name | metagenome_name | kegg_module | module_name | module_class | module_category | module_subcategory | module_definition | kofam_hits_in_module | gene_caller_ids_in_module | module_is_complete | module_completeness
---|---|---|---|---|---|---|---|---|---|---|---|---|
0 | sample_101 | sample_101 | M00001 | Glycolysis (Embden... | Pathway modules | Carbohydrate metabolism | Central carbohydrate metabolism | """(K00844,K12407,K00845,... | K21071,K00134,... | 33382,34494, ... | True | 1.0
...|...|...|...|...|...|...|...|...|...|...|...|...|
5301 | sample_146 | sample_146 | M00618 | Acetogen | Signature modules | Module set | Metabolic capacity | """M00377 M00579""" | K00297,K00625,... | 45332,44589, ... | True | 0.9285714285714286

**Example groups.txt**

sample | group
---|---
sample_1 | R
sample_2	| H
sample_3	| R
sample_4	| N
... | ...

**Using a different column for sample names**
The sample names in `sample` column of the `groups.txt` must match those in a corresponding column in the `modules.txt` - this column is `db_name` by default (because we assume most people will be using metabolism output from multiple samples, and metabolism multi-mode output typically includes a `db_name` column). However, you can specify a different column that has sample names using the `--sample-header` parameter, like so:
```
anvi-compute-enrichment-scores -M modules.txt -G groups.txt -o enriched_modules.txt --sample-header metagenome_name
```

**The role of the module completion threshold parameter**
It is easy to know which functions are present in which genomes because, well, if you have an annotation for it, it's there. It is less straightforward for metabolic modules because a sample may have only some of the KOs in the module definition. 

'Modules' mode output from `anvi-estimate-metabolism` includes a line for every module in every sample, so how do we tell which sample actually includes a module or not? The answer is the module completeness score (the % of KOs in the module definition that have a hit in the sample): if the module completeness score is over the module completion threshold, then we consider that module to be 'present' in that particular sample, and we include it in the input for the enrichment script. 

The threshold by default is 0.75 (which is the totally arbitrary number I have also been using as the default for completion in `anvi-estimate-metabolism`), but can be changed using the `--module-completion-threshold` parameter:

```
anvi-compute-enrichment-scores -M modules.txt -G groups.txt -o enriched_modules.txt --module-completion-threshold 0.6
```

### CODE REVIEW

**Comment by adw96** [comment]
Very, very cool, @ivagljiva ! Thank you so much for doing this! 

I'm so glad that the script and analysis I proposed for pangenomics have been useful. That said, I developed it with a very specific use case (pangenomic comparisons) in mind. In different settings, a different approach to modelling might be more appropriate. Would you be interested in having a (eg zoom) call to talk about the metabolism case? It may be that there is additional data could be useful to improve inference here -- `module_completeness` data, for example. Unfortunately I'm not familiar with `anvi-estimate-metabolism` (YET) so perhaps we could chat about that. 

I shouldn't say this publicly, but I can be broadly available to chat with you and @meren about this 😸 Feel free to reach out to me by Slack to set up a time.

**Comment by ivagljiva** [comment]
## Update: Functional enrichment for pangenomes now uses generalized enrichment script 

No matter whether you run pangenome functional enrichment through `anvi-compute-enrichment-scores` or the older `anvi-get-enriched-functions-per-pan-group`, now this will use the generalized `anvi-script-enrichment-stats` rather than the original `anvi-script-run-functional-enrichment-stats`. The only practical effect of this change is that the `function_accession` header in the input/output files for the script is now `accession`. 

Please note that `anvi-script-run-functional-enrichment-stats` is still around. I could remove it entirely from our sandbox, but for now I decided to leave it because we do not know if anyone is using it independently. 


## New Addition: Function enrichment for groups of genomes

I have extended `anvi-script-enrichment-stats` so that people can now look for enriched functions in groups of genomes. This works in essentially the same way as functional enrichment for pangenomes, in that we count up how many times each function occurs in each group, generate an input file containing the proportions, and then run the generalized enrichment script on that input file. 

**Input**
We accept external genomes files and internal genomes files (one or both of these) as input. These files must have a 'group' column that specifies which group each genome belongs to. For example, here is an internal genomes file with the group column:

name | contigs_db_path | profile_db_path | collection_id | bin_id | group
----|----|----|----|----|----
meta_bin_1 | INFANT-GUT-TUTORIAL/CONTIGS.db | INFANT-GUT-TUTORIAL/PROFILE.db | TEST_6_BINS | Bin_1 | A
meta_bin_2 | INFANT-GUT-TUTORIAL/CONTIGS.db | INFANT-GUT-TUTORIAL/PROFILE.db | TEST_6_BINS | Bin_2 | B
meta_bin_3 | INFANT-GUT-TUTORIAL/CONTIGS.db | INFANT-GUT-TUTORIAL/PROFILE.db | TEST_6_BINS | Bin_3 | A
meta_bin_4 | INFANT-GUT-TUTORIAL/CONTIGS.db | INFANT-GUT-TUTORIAL/PROFILE.db | TEST_6_BINS | Bin_4 |
meta_bin_5 | INFANT-GUT-TUTORIAL/CONTIGS.db | INFANT-GUT-TUTORIAL/PROFILE.db | TEST_6_BINS | Bin_5 | A
meta_bin_6 | INFANT-GUT-TUTORIAL/CONTIGS.db | INFANT-GUT-TUTORIAL/PROFILE.db | TEST_6_BINS | Bin_6 | B

Note that empty values are allowed in the 'group' column (see `meta_bin_4` in the example above) as long as not _every_ genome has an empty group. Genomes with empty group column values will be considered 'ungrouped' and can be excluded from the enrichment analysis using the `--exclude-ungrouped` flag, though by default they will be included and considered as their own group (called 'None' - though please note that if you literally put the string 'None' in the 'group' column those genomes will be considered to be in the same group as those with empty groups, and excluding them will no longer work).

**Usage**

In addition to the input file(s) you must provide an annotation source, just as in the pangenome functional enrichment case. Use `--list-annotation-sources` to see which ones are available. You must also provide an output file with `-o`. 

Here is an example in which we run on both internal and external genomes using source 'KOfam':
```
anvi-compute-enrichment-scores -e ext_genomes_w_groups.txt -i int_genomes_w_groups.txt -o enrichment.txt --annotation-source KOfam
```

And just like the pangenome case, you can also elect to save a table of function annotation counts, using the `--functional-occurrence-table-output` parameter. 

**Testing**

I want to test this on the [Prochlorococcus metapangenome](https://merenlab.org/data/prochlorococcus-metapangenome/) to verify that the enriched function results are as expected. However, I would like some input on this as I am not sure I am going about it the right way. See below for details.

_Preparing data and computing enrichment for groups of Prochlorococcus isolate_
I downloaded the datapack for all the Prochlorococcus isolate genomes from [here](https://figshare.com/articles/dataset/Anvi_o_profile_for_Prochlorococcus_isolates_across_TARA_Oceans_metagenomes/5447224). I migrated the databases, imported the collection into the profile database, and split the project to get one contigs DB for each isolate genome:
```
tar -xvzf ANVIOPROFILESFORPROCHLOROCOCCUSISOLATESACROSSTARA.tar.gz
cd ANVIO-PROFILES-FOR-PROCHLOROCOCCUS-ISOLATES-ACROSS-TARA/
anvi-migrate --migrate-dbs-safely *.db   # what an amazing script
anvi-import-collection anvio-collection-for-pro-isolates.txt -c CONTIGS.db -p PROFILE.db -C GENOMES
anvi-split -p PROFILE.db -c CONTIGS.db -C GENOMES -o SPLIT_PROCHLOROCOCCUS
```
Then I made an external genomes file, organizing each of the isolate genomes into either the HL group or the LL group, following the guidance of the [Prochlorococcus paper](https://peerj.com/articles/4320/). Here it is.
[prochloro_ext_genomes.txt](https://github.com/merenlab/anvio/files/5662091/prochloro_ext_genomes.txt)

I ran the enrichment script using COG functions as the annotation source, because that is the source Alon used when doing this in the [pangenomics tutorial](https://merenlab.org/2016/11/08/pangenomics-v2/#making-sense-of-functions-in-your-pangenome).
```
anvi-compute-enrichment-scores -e prochloro_ext_genomes.txt -o prochloro_enrichment.txt -F prochloro_func_occurrence.txt --annotation-source COG14_FUNCTION
```
Here are the output files:
[prochloro_enrichment.txt](https://github.com/merenlab/anvio/files/5662465/prochloro_enrichment.txt)
[prochloro_func_occurrence.txt](https://github.com/merenlab/anvio/files/5662466/prochloro_func_occurrence.txt)

_**This is the part where I would like some input**_
I need to compare this output to something to see whether we've found the right enriched functions. 
I think I can compare this output to the enriched functions we get on the same genomes via pangenome functional enrichment (as described in the pangenomics tutorial), because I believe they should be somewhat similar. Here is my reasoning.

The difference between the two approaches is that pangenome functional enrichment considers occurrence of functions at the level of gene clusters, while the genome-groups approach considers occurrence of functions at the level of individual annotated genes. That is, with a pangenome the steps for computing enrichment are like this:
1. for each gene cluster, set the most frequently occurring functional annotation (from the provided annotation source) to be THE function for that GC.
2. count how many times each function occurs in each genome by summing the count of gene calls from that genome in all GCs that have this function.
3. make the occurrence binary: either the function occurs in the genome (count > 0) or it doesn't (count = 0)
4. count the proportion of genomes in each category that have the function as 'occurring', and run the enrichment analysis based on these proportions to find which categories have the function more than expected.

With the genome-groups approach, steps 1 and 2 are different, and look like this:
1. take all gene calls with an annotation (from the provided annotation source)
2. count how many times each function occurs in each genome by summing the count of gene calls from that genome that have this function. Accessions with the same annotation are considered to be the same function. 

This means that the pangenome enrichment will count only functional annotations that are majority in the gene clusters, while genome-groups enrichment will count all functional annotations. Furthermore, since pangenome enrichment converts the non-majority annotations into the majority function, the majority function may have higher counts than the number of actual annotations of that function. Using presence/absence rather than raw counts for the final proportion calculation helps, but we will still get different enrichment results even with the same genomes and the same annotation source. 

However, if a function is truly enriched, it _should_ end up as the majority in its gene clusters, and thus be counted by the pangenome method. It would have a high enrichment score if lots of genomes in a category had genes in those GCs. Likewise, in the genome-groups method the function would be annotated (so of course it would be counted), and it would have a high enrichment score if it was annotated in a lot of genomes in a category. I think it is likely that an enriched function in the genome-groups case would also be enriched in the pangenome case. 

The vice versa is a bit less likely - to see why, consider this example: GC1 has 5 genes with annotations 'A', 'A', 'A', 'B', 'C'. Each gene is from a different genome in the SAME category, "CAT", and those genes are the only ones with annotations to 'A', 'B', or 'C' in the entire dataset. Since 'A' is the majority annotation, we take consider all genes in GC1 to be 'A', which means that all 5 genomes in the category "CAT" have 'A' occurring. Supposing that 'A' does not occur in any genomes in other categories, 'A' will have a high enrichment score and be associated with category "CAT". 'B' and 'C' will not appear in the enrichment results. But now suppose we do the analysis using the genome-groups method, in which all 5 genomes are in the same group "CAT". This time, we consider each functional annotation separately so 'A' is only in 3/5 genomes in "CAT" while 'B' and 'C' each occur in 1/5 genomes in "CAT". All three functions will appear in the enrichment results, but 'A' will have a lower enrichment score than in the pangenome case. 

In conclusion, I think that if I compare the genome-groups enrichment file to the pangenome enrichment file for the same genomes/groups/annotation source, I should see that functions with high enrichment scores from the genome-groups case appear with high enrichment scores in the pangenome case, but that functions with high enrichment scores in the pangenome case are not necessarily very enriched in the genome-groups case.

@meren Does this make sense? Is there a better way to confirm that our results are as expected?

**Comment by meren** [comment]
This is gorgeous, @ivagljiva.

> We accept external genomes files and internal genomes files (one or both of these) as input. These files must have a 'group' column that specifies which group each genome belongs to.

So we essentially changed the standards of `external-genomes` and `internal-genomes` artifacts. Perhaps updating those artifacts to mention that they also may include an optional `group` column would be a good idea.

Side note, I recently added this artifact which also has an optional `group` column that is relevant depending on the context:

https://merenlab.org/software/anvio/help/main/artifacts/samples-txt/

To make sure everyone can benefit from a single sanity check point I also implemented a utils function (`get_samples_txt_file_as_dict`) to read this artifact type. Not now, but at some point we should probably think about how to best centralize the sanity checking these files so their standards are clear everywhere.

> Note that empty values are allowed in the 'group' column (see meta_bin_4 in the example above) as long as not every genome has an empty group.

I was going to ask for this, but clearly you are on top of everything :)

> And just like the pangenome case, you can also elect to save a table of function annotation counts, using the --functional-occurrence-table-output parameter.

<3

> The difference between the two approaches is that pangenome functional enrichment considers occurrence of functions at the level of gene clusters, while the genome-groups approach considers occurrence of functions at the level of individual annotated gene

Are you sure? I think pangenome also uses individual genes and not gene clusters. You *can* use gene cluster identity as function, but I think the approach essentially uses genes and their functions learned from the genome storage. I am sorry if I am missing something, which is very likely since you have a much better understanding of this now :)

> In conclusion, I think that if I compare the genome-groups enrichment file to the pangenome enrichment file for the same genomes/groups/annotation source, I should see that functions with high enrichment scores from the genome-groups case appear with high enrichment scores in the pangenome case

Regardless, I do agree with this.

**Comment by ivagljiva** [comment]
Thanks @meren 😊 

> Perhaps updating those artifacts to mention that they also may include an optional group column would be a good idea.

I am on it. Documentation is a great evening activity :) 

> Are you sure? I think pangenome also uses individual genes and not gene clusters.

I think so.... Actually I think we are both right, we are just talking about it in different ways. The pangenome case uses individual genes, but we get the functional summary using the `dbops.py` function `get_gene_cluster_function_summary`, which results in each of the genes in a gene cluster being assigned the same function - (the most common function in the gene cluster). That is what I meant, sorry if I didn't put it the right way. But if this still does not sound right to you, please let me know :) 

> Regardless, I do agree with this.

Great, then I will go forward with the comparison to the pangenome results and put what I find here. If it looks good, I hope we can merge this for v7 :)

**Comment by ivagljiva** [comment]
@meren, I added a help page for `anvi-script-enrichment-stats` (see last commit), but it is not getting detected when I run `anvi-script-gen-help-pages`:

image

Do you know what might be the problem?

**Comment by meren** [comment]
Hey @ivagljiva, that is an R script and will not be a part of our documentation directly :/ but no one should be using that script directly since the actual interface to that is `anvi-compute-enrichment-scores` (which I now feel like we should call `anvi-compute-functional-enrichment` :)).

In the help menu of this program we can mention that its purpose is to serve as an interface to `anvi-script-enrichment-stats`, which does this and that. Does this make sense as a solution?

**Comment by ivagljiva** [comment]
This program is now called `anvi-compute-functional-enrichment` :) Documentation has been updated, but the comments above have the old name so if you are testing those out, please remember to change the program name.

**Comment by ivagljiva** [comment]
## Comparison to pangenome results

I ran functional enrichment for the Prochlorococcus pangenome by following the pangenomics tutorial. 

As a reminder, this is what we should expect:
> functions with high enrichment scores from the genome-groups case [should] appear with high enrichment scores in the pangenome case

And this is indeed what I am seeing :) For example, here are the top enriched functions in the genome-groups output:
```
$ cut -f 1-5 ANVIO-PROFILES-FOR-PROCHLOROCOCCUS-ISOLATES-ACROSS-TARA/prochloro_enrichment.txt | head -n 20
COG14_FUNCTION	enrichment_score	unadjusted_p_value	adjusted_q_value	associated_groups
ATP-dependent DNA ligase	31.00002278725131	2.5802540103336655e-8	2.041267617063986e-6	HL
Hydroxymethylpyrimidine/phosphomethylpyrimidine kinase	31.00002278725131	2.5802540103336655e-8	2.041267617063986e-6	HL
N-acetylglutamate synthase or related acetyltransferase, GNAT family	31.00002278725131	2.5802540103336655e-8	2.041267617063986e-6	HL
RecB family exonuclease	31.00002278725131	2.5802540103336655e-8	2.041267617063986e-6	HL
Redox-sensitive bicupin YhaK, pirin superfamily	31.00002278725131	2.5802540103336655e-8	2.041267617063986e-6	HL
Thiaminase	31.00002278725131	2.5802540103336655e-8	2.041267617063986e-6	HL
1,6-Anhydro-N-acetylmuramate kinase	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
3-polyprenyl-4-hydroxybenzoate decarboxylase	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
Exonuclease VII small subunit	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
Exonuclease VII, large subunit	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
N-acetylglucosamine-6-phosphate deacetylase	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
N-acetylmuramic acid 6-phosphate (MurNAc-6-P) etherase	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
Na+/proline symporter	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
Periplasmic beta-glucosidase and related glycosidases	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
Proteasome lid subunit RPN8/RPN11, contains Jab1/MPN domain metalloenzyme (JAMM) motif	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
Ser/Thr protein kinase RdoA involved in Cpx stress response, MazF antagonist	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
Single-stranded DNA-specific exonuclease, DHH superfamily, may be involved in archaeal DNA replication intiation	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
Uncharacterized integral membrane protein	31.00002278725129	2.5802540103336903e-8	2.041267617063986e-6	LL
3-hydroxyisobutyrate dehydrogenase or related beta-hydroxyacid dehydrogenase	26.9958340715296	2.0389443023494397e-7	1.5281350981818958e-5	LL
```

Most of these functions end up at the top of the pangenomics output as well, albeit with a different ranking:
```
$ cut -f 1-5 Prochlorococcus_31_genomes/PROCHLORO-PAN-enriched-functions-light.txt | head -n 20
COG14_FUNCTION	enrichment_score	unadjusted_p_value	adjusted_q_value	associated_groups
Thiaminase	31.00002278725131	2.5802540103336655e-8	1.6322911239284867e-6	HL
RecB family exonuclease	31.00002278725131	2.5802540103336655e-8	1.6322911239284867e-6	HL
Redox-sensitive bicupin YhaK, pirin superfamily	31.00002278725131	2.5802540103336655e-8	1.6322911239284867e-6	HL
ATP-dependent DNA ligase	31.00002278725131	2.5802540103336655e-8	1.6322911239284867e-6	HL
N-acetylglutamate synthase or related acetyltransferase, GNAT family	31.00002278725131	2.5802540103336655e-8	1.6322911239284867e-6	HL
Chromatin remodeling complex protein RSC6, contains SWIB domain	31.00002278725131	2.5802540103336655e-8	1.6322911239284867e-6	HL
Hydroxymethylpyrimidine/phosphomethylpyrimidine kinase	31.00002278725131	2.5802540103336655e-8	1.6322911239284867e-6	HL
Predicted NAD/FAD-dependent oxidoreductase	31.00002278725131	2.5802540103336655e-8	1.6322911239284867e-6	HL
Exonuclease VII small subunit	31.00002278725129	2.5802540103336903e-8	1.6322911239284867e-6	LL
3-hydroxyisobutyrate dehydrogenase or related beta-hydroxyacid dehydrogenase	31.00002278725129	2.5802540103336903e-8	1.6322911239284867e-6	LL
1,6-Anhydro-N-acetylmuramate kinase	31.00002278725129	2.5802540103336903e-8	1.6322911239284867e-6	LL
8-oxo-dGTP pyrophosphatase MutT and related house-cleaning NTP pyrophosphohydrolases, NUDIX family!!!Adenine-specific DNA glycosylase, acts on AG and A-oxoG pairs	31.00002278725129	2.5802540103336903e-8	1.6322911239284867e-6	LL
3-polyprenyl-4-hydroxybenzoate decarboxylase	31.00002278725129	2.5802540103336903e-8	1.6322911239284867e-6	LL
Ser/Thr protein kinase RdoA involved in Cpx stress response, MazF antagonist	31.00002278725129	2.5802540103336903e-8	1.6322911239284867e-6	LL
N-acetylglucosamine-6-phosphate deacetylase	31.00002278725129	2.5802540103336903e-8	1.6322911239284867e-6	LL
```

There were only 4 functions in the top 19 of the genomes-group that we don't see in the pangenome top 19 output: those are 'Exonuclease VII, large subunit', 'N-acetylmuramic acid 6-phosphate (MurNAc-6-P) etherase', 'Periplasmic beta-glucosidase and related glycosidases', and 'Uncharacterized integral membrane protein'. These functions are still in the pangenome output of course, just farther down in the ranking. 

Most of the differences in ranking seem to be due to having different q values.

So I think this is good to go! @adw96, once you take a look at this and approve it, I think we can merge it. :)

**Comment by adw96** [comment]
@ivagljiva 🍄 -- you are a total pro. I think your sanity checks were very sensible and this is a good generalisation of the functional enrichment script to a broader class of inputs. LGTM.

@meren  -- as you know, the StatDivLab is in the process of proposing and refining a statistical inference procedure for testing module enrichment. The timeframe for a fabulous, state-of-the-art procedure could be Summer 2021. However, we may be able to come up with a straightforward modification of off-the-shelf procedures sooner.

@ivagljiva 's current implementation is sensible, but at some point we are planning to change the model. There are also choices to be made about **what to model** in the modules case that didn't exist in the functions case. For example, are we interested in assessing if modules are more complete on average, or if they have a higher probability of being fully complete? In my view, these are different questions that we could answer -- both are valid and warrant discussion.

Short version of my question for @meren  -- should we make this available now to make *something* accessible to the community sooner (even if it is subject to change)? Or should we wait to finalise the modelling question and release an off-the-shelf method at a later (but soonish) date?

**Comment by meren** [comment]
Hey @adw96, great questions :)

> The timeframe for a fabulous, state-of-the-art procedure could be Summer 2021. However, we may be able to come up with a straightforward modification of off-the-shelf procedures sooner.

I think what @ivagljiva's `anvi-estimate-metabolism` is already reporting, and thanks to the more generalized functional enrichment procedure, we are addressing the most basic needs of working with metabolisms in an integrated fashion. Despite the vast room for improvement, there is no immediate pressure, I think. Which makes me believe it is worth working with a timetable that will make you and others in the StatDivLab to be happy with the final product (rather than spending valuable effort to put together a quick solution which you will likely declare as obsolete soon after :)).

> There are also choices to be made about what to model in the modules case that didn't exist in the functions case. For example, are we interested in assessing if modules are more complete on average, or if they have a higher probability of being fully complete? In my view, these are different questions that we could answer -- both are valid and warrant discussion.

Yes, these are essential considerations as we go from individual genes/functions spaces to metabolic pathways realm. I think the ability to report the degree of completeness of a module with a degree of confidence is an essential need. So I guess my preliminary answer to your question is .. both? :) But since we have access to much more well-structured information compared to off-the-shelf tools for metabolic reconstruction solutions, I think there are other opportunities we can exploit. So we certainly should have discussions! :)


Best,
